src/types/Stream.d.ts

```diff
@@ -0,0 +1,69 @@
+import Client from './Client'
+import { ExprArg } from './query'
+import { values } from './values'
+
+export interface Subscription {
+  on: <T extends keyof SubscriptionEventHandlers>(
+    type: T,
+    callback: SubscriptionEventHandlers[T]
+  ) => this
+  start: () => this
+  close: () => void
+}
+
+type SubscriptionEventHandlers = {
+  start: (type: 'start', txn: values.FaunaTime, event: values.FaunaTime) => void
+  error: (
+    type: 'error',
+    txn: values.FaunaTime,
+    event: {
+      code: string
+      description: string
+    }
+  ) => void
+  version: (
+    type: 'version',
+    txn: values.FaunaTime,
+    event: {
+      action: 'create' | 'update' | 'delete'
+      document: {
+        ref: values.Ref
+        ts: values.FaunaTime
+        data: object
+      }
+      diff: {
+        ref: values.Ref
+        ts: values.FaunaTime
+        data: object
+      }
+      prev: {
+        ref: values.Ref
+        ts: values.FaunaTime
+        data?: object
+      }
+    }
```

Comment:
I appreciate the careful typing here but, I think this are better off as just `object` as their rendering are highly dynamic and dependent on the document being streamed. For instance, data is never guaranteed to be present. Once we enable schema documents, they will have other properties, like `name`, `roles`, etc. Trying to cover all possibilities at this level will be quite tedious and error prone.

Reply:
This is the same for other `document` fields in this type annotation.

Reply:
If I can provide some user feedback. I would much rather have incomplete/error-prone types over simply `any` or `object`. TS types provide not only type-safety for refactoring but also serve as a main source for documentation when learning/exploring new parts of the api.

Reply:
As a user, can you work around incomplete types in TS? If so, I'd say mark data as optional and leave it as is.